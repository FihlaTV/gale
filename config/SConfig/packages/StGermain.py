import os
import SConfig

class StGermain(SConfig.Package):
    def __init__(self, env, options):
        SConfig.Package.__init__(self, env, options)
        self.require(SConfig.packages.cmath)
        self.require(SConfig.packages.libXML2)
        self.require(SConfig.packages.MPI)
        self.base_patterns = ['StGermain*']
        self.headers = [[os.path.join('StGermain', 'StGermain.h')]]
        self.libraries = [['StGermain']]
        self.symbols = [(['StGermain_Init', 'StGermain_Finalise'], '')]
        self.symbol_setup = 'MPI_Init(&argc, &argv);'
        self.symbol_teardown = 'MPI_Finalize();'
        self.symbol_calls = ['%s(&argc, &argv);', '%s();']
        self.require_shared = True
        self.use_rpath = True

    def get_run_error_message(self, console):
        if len(console):
            return 'Incompatible libraries, check \'config.log\'.'
