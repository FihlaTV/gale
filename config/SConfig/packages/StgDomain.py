import os
import SConfig

class StgDomain(SConfig.Package):
    def __init__(self, env, options):
        SConfig.Package.__init__(self, env, options)
        self.setup_search_defaults()
        self.setup_options()
        self.dependencies = [SConfig.packages.StGermain]
        self.base_patterns = ['StgDomain*']
        self.headers = [os.path.join('StgDomain', 'StgDomain.h')]
        self.dependency_headers = [os.path.join('StGermain', 'StGermain.h')]
        self.libraries = [['StgDomain']]
        self.symbols = [(['StgDomain_Init', 'StgDomain_Finalise'], '')]
        self.symbol_setup = '''MPI_Init(&argc, &argv);
StGermain_Init(&argc, &argv);
'''
        self.symbol_teardown = '''StGermain_Finalise();
MPI_Finalize();
'''
        self.symbol_calls = ['%s(&argc, &argv);', '%s();']
        self.require_shared = True
        self.use_rpath = True

    def get_run_error_message(self, console):
        if len(console):
            return 'Incompatible libraries, check \'config.log\'.'
