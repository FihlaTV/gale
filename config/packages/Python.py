import os, glob
import config
import config.utils as utils

class Python(config.Package):

    def __init__(self, ctx, **kw):
        config.Package.__init__(self, ctx, **kw)
        self.version = None

    def setup_libraries(self):
        self.add_library_set(["Python.h"], ["python"])

    def gen_inc_exts(self, inc_dirs):
        ext_map = []
        inc_d = inc_dirs[0]
        for d in glob.glob(os.path.join(inc_d, "python*")):
            if not os.path.isdir(d):
                continue
            d = os.path.basename(d)
            try:
                n = map(int, d[6:].split("."))
                ext_map.append((n, os.path.basename(d)))
            except:
                pass
        if ext_map:
            ext_map.sort(lambda x,y: y[0] < x[0] and -1 or y[0] > x[0] and 1 or 0)
            self.version = ext_map[0][0]
            yield ext_map[0][1]
        for e in config.Package.gen_inc_exts(self, inc_dirs):
            yield e

    def gen_lib_cands(self, dirs):
        if self.version is not None:
            version = ".".join([str(i) for i in self.version])
            for env in self._lib_sets:
                new_libs = []
                for l in env["libs"]:
                    if l == "python":
                        new_libs.append(l + version)
                    else:
                        new_libs.append(l)
                env["libs"] = new_libs
        for c in config.Package.gen_lib_cands(self, dirs):
            yield c

    source_code = {"c": """#include <stdlib.h>
#include <Python.h>
int main(int argc, char** argv) {
  return EXIT_SUCCESS;
}
"""}
