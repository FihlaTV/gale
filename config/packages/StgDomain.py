import os
import SConfig
import SCons.Script

class StgDomain(SConfig.Package):
    def __init__(self, env, options):
        SConfig.Package.__init__(self, env, options)
        self.checks = [self.setup_environment]
        self.dependencies = [SConfig.packages.cmath,
                             SConfig.packages.libXML2,
                             SConfig.packages.MPI,
                             SConfig.packages.SVNRevision,
                             SConfig.packages.BlasLapack,
                             SConfig.packages.StGermain]
        if self.opts:
            self.opts.AddOptions(
                SCons.Script.BoolOption('debug',
                                        'Enable debugging', 1),
                SCons.Script.BoolOption('staticLibraries',
                                        'Build static libraries only', 0),
                ('buildPath', 'Temporary build path', 'build'))

    def setup_environment(self):
        # Setup the build path.
        if not os.path.isabs(self.env['buildPath']):
            self.env['buildPath'] = '#' + self.env['buildPath']

        # Setup LIB_DIR.
        lib_dir = os.path.join(self.env['buildPath'], 'lib')
        abs_lib_dir = self.env.Dir(lib_dir).abspath
        self.env.AppendUnique(CPPDEFINES=[('LIB_DIR',
                                           self.env['ESCAPE']('"' + abs_lib_dir + '"'))])
        self.env.PrependUnique(LIBPATH=[lib_dir])

        # Setup the RPATH.
        self.env.PrependUnique(RPATH=[self.env.Dir(abs_lib_dir).abspath])

        # Setup the module extension.
        ext = self.env['ESCAPE']('"' + self.env['SHLIBSUFFIX'][1:] + '"')
        self.env.Append(CPPDEFINES=[('MODULE_EXT', ext)])

        # Setup the include paths.
        inc_path = os.path.join(self.env['buildPath'], 'include')
        self.env.AppendUnique(CPPPATH=[inc_path])
        self.env.AppendUnique(CPPPATH=[os.path.join(inc_path, 'StgDomain')])

        # Setup debugging.
        if self.env['debug']:
            self.env.MergeFlags('-g')

        # Setup 64 bit builds.
        if platform.architecture()[0].find('64') != -1:
            self.env.MergeFlags('-m64')

        return [1, '', '']
