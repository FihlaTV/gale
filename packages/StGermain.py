import os
import config

class StGermain(config.Package):

    def __init__(self, ctx, **kw):
        config.Package.__init__(self, ctx, **kw)
        self.inc_exts = ["StGermain"]

    def setup_dependencies(self):
        config.Package.setup_dependencies(self)
        self.add_dependency(config.packages.MPI, required=True, combine=True)
        self.add_dependency(config.packages.libxml2, required=True, combine=True)

    def setup_libraries(self):
        self.add_library_set([os.path.join("StGermain", "StGermain.h")], ["StGermain"])

    def setup_options(self):
        config.Package.setup_options(self)
        self.options.add_option(
            config.utils.options.Option("stg_dir", "--stg-dir",
                                        pref_sep="=", help=self.help["stg_dir"]))

    def process_options(self):
        config.Package.process_options(self)
        base_dir = self.get_option("stg_dir", None)
        if base_dir is not None:
            self.forced_base_dirs = [base_dir]

    def make_help_dict(self):
        config.Package.make_help_dict(self)
        self.help["stg_dir"] = "Specify the base location for all StG* related projects."

    source_code = {"c": """#include <stdlib.h>
#include <StGermain/StGermain.h>
int main( int argc, char** argv ) {
  return EXIT_SUCCESS;
}
"""}
