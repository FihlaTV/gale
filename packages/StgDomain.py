import os
import config

class StgDomain(config.Package):

    def __init__(self, ctx, **kw):
        config.Package.__init__(self, ctx, **kw)
        self.inc_exts = ["StgDomain"]

    def setup_dependencies(self):
        config.Package.setup_dependencies(self)
        self.stg = self.add_dependency(config.packages.StGermain, required=True, combine=True)
        self.add_dependency(config.packages.BlasLapack, required=True, combine=True)

    def setup_libraries(self):
        self.add_library_set([os.path.join("StgDomain", "StgDomain.h")], [])

    def process_options(self):
        config.Package.process_options(self)
        base_dir = self.stg.get_option("stg_dir", None)
        if base_dir is not None:
            self.forced_base_dirs = [base_dir]

    source_code = {"c": """#include <stdlib.h>
#include <StGermain/StGermain.h>
#include <StgDomain/StgDomain.h>
int main( int argc, char** argv ) {
  return EXIT_SUCCESS;
}
"""}
