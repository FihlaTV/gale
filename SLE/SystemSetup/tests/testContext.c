/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003-2006, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street,
**	Melbourne, 3053, Australia.
**
** Primary Contributing Organisations:
**	Victorian Partnership for Advanced Computing Ltd, Computational Software Development - http://csd.vpac.org
**	Australian Computational Earth Systems Simulator - http://www.access.edu.au
**	Monash Cluster Computing - http://www.mcc.monash.edu.au
**	Computational Infrastructure for Geodynamics - http://www.geodynamics.org
**
** Contributors:
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Robert Turnbull, Research Assistant, Monash University. (robert.turnbull@sci.monash.edu.au)
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	David May, PhD Student, Monash University (david.may@sci.monash.edu.au)
**	Louis Moresi, Associate Professor, Monash University. (louis.moresi@sci.monash.edu.au)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**	Julian Giordani, Research Assistant, Monash University. (julian.giordani@sci.monash.edu.au)
**	Vincent Lemiale, Postdoctoral Fellow, Monash University. (vincent.lemiale@sci.monash.edu.au)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** $Id: testContext.c 964 2007-10-11 08:03:06Z SteveQuenette $
**
** Purpose: primarily in this test we're going to check the timestep
**	limiting works.
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#ifdef HAVE_PYTHON
	#include <Python.h>
#endif
#include <mpi.h>
#include <StGermain/StGermain.h>
#include <StgDomain/StgDomain.h>
#include "StgFEM/Discretisation/Discretisation.h"
//#include "StgFEM/SLE/LinearAlgebra/LinearAlgebra.h"
#include "StgFEM/SLE/SystemSetup/SystemSetup.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct _Element {
	__FiniteElement_Element
};

struct _Particle {
	__IntegrationPoint
};


double testCalcDtFunc( FiniteElementContext* context, FiniteElementContext* contextArg2 ) {
	if ( context->timeStep == 1 ) {
		return 1.0;
	}
	else {
		return 1.5 * context->dt;
	}
}

int main( int argc, char* argv[] ) 
{
	/* StGermain standard bits & pieces */
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	Dictionary*			dictionary;
	Dictionary*			componentDict;
	XML_IO_Handler*			ioHandler;
	Stream*                         stream;
	
	/* context */
	FiniteElementContext*		context = NULL;
	
	/* Initialise PETSc, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );
	
	StGermain_Init( &argc, &argv );
	StgFEM_Discretisation_Init( &argc, &argv );
	//StgFEM_SLE_LinearAlgebra_Init( &argc, &argv );
	StgFEM_SLE_SystemSetup_Init( &argc, &argv );
	MPI_Barrier( CommWorld ); /* Ensures copyright info always come first in output */
	
	stream = Journal_Register( Info_Type, "test" );
	Stream_SetPrintingRank( stream, 0 );
	
	/* Create the application's dictionary */
	dictionary = Dictionary_New();

	Dictionary_Add( dictionary, "outputPath", Dictionary_Entry_Value_FromString( "output" ) );

	/* Read input */
	ioHandler = XML_IO_Handler_New();
	IO_Handler_ReadAllFromCommandLine( ioHandler, argc, argv, dictionary );

	Journal_ReadFromDictionary( dictionary );

	/* Construction phase ----------------------------------------------------------------------------------------------*/
	context = FiniteElementContext_New( "context", 0, 0, CommWorld, dictionary );
	
	Stream_SetPrintingRank( context->info, 0 );

	componentDict = Dictionary_GetDictionary( dictionary, "components" );

	if ( componentDict == NULL ) {
		componentDict = Dictionary_New();
	}
	context->CF = Stg_ComponentFactory_New( dictionary, componentDict, context->register_Register );
	
	/* This is where we'd normally construct components if it was real main.
		instead, we'll just set the dt function so we can test it */
	EP_AppendClassHook( context->calcDtEP, testCalcDtFunc, context );
	
	if( rank == 0 ) 
		Context_PrintConcise( context, context->verbose );

	if ( True == Dictionary_GetBool_WithDefault( dictionary, "showJournalStatus", False ) ) {
		Journal_PrintConcise();	
	}	

	/* Building phase ---------------------------------------------------------------------------------------------------*/
	Stg_Component_Build( context, 0 /* dummy */, False );
	
	/* Initialisaton phase ----------------------------------------------------------------------------------------------*/
	Stg_Component_Initialise( context, 0 /* dummy */, False );
	
	/* Run (Solve) phase ------------------------------------------------------------------------------------------------*/
	AbstractContext_Dump( context );

	context->maxTimeSteps = 10;

	Journal_Printf( stream, "Running with no timestep braking, using  "
		"dt that increases 50%% each step:\n" );

	Stg_Component_Execute( context, 0 /* dummy */, False );
	context->currentTime=0;
	context->dt = 0;

	Journal_Printf( stream, "\nTurning on timestep braking, at default "
		"level, running again:\n" );
	context->limitTimeStepIncreaseRate = True;
	Stg_Component_Execute( context, 0 /* dummy */, True );
	context->currentTime=0;
	context->dt = 0;

	Journal_Printf( stream, "\nTurning on timestep braking, at 80%% "
		"level, running again - expect same as original:\n" );
	context->maxTimeStepIncreasePercentage = 80;
	Stg_Component_Execute( context, 0 /* dummy */, True );

	/* Destruct phase ---------------------------------------------------------------------------------------------------*/
	Stg_Component_Destroy( context, 0 /* dummy */, False );
	Stg_Class_Delete( context );
	Stg_Class_Delete( dictionary );

	StgFEM_SLE_SystemSetup_Finalise();
	//StgFEM_SLE_LinearAlgebra_Finalise();
	StgFEM_Discretisation_Finalise();
	StGermain_Finalise();
		
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
