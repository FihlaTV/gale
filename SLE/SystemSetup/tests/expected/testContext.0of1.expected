StGermain Framework revision 3885. Copyright (C) 2003-2005 VPAC.
StGermain Discretisation Library revision 3885. Copyright (C) 2003-2005 VPAC.
StGermain FEM Discretisation Framework revision 662. Copyright (C) 2003-2005 VPAC.
StGermain PETSc-LinearAlgebra Interface revision 662. Copyright (C) 2003-2005 VPAC.
Running with no timestep braking, using  dt that increases 50% each step:
Run until 10 timeSteps have been run
TimeStep = 1, Start time = 0 + 0 prev timeStep dt
TimeStep = 2, Start time = 0 + 1 prev timeStep dt
TimeStep = 3, Start time = 1 + 1.5 prev timeStep dt
TimeStep = 4, Start time = 2.5 + 2.25 prev timeStep dt
TimeStep = 5, Start time = 4.75 + 3.375 prev timeStep dt
TimeStep = 6, Start time = 8.125 + 5.0625 prev timeStep dt
TimeStep = 7, Start time = 13.1875 + 7.59375 prev timeStep dt
TimeStep = 8, Start time = 20.7812 + 11.3906 prev timeStep dt
TimeStep = 9, Start time = 32.1719 + 17.0859 prev timeStep dt
TimeStep = 10, Start time = 49.2578 + 25.6289 prev timeStep dt

Turning on timestep braking, at default level, running again:
Run until 10 timeSteps have been run
TimeStep = 1, Start time = 0 + 0 prev timeStep dt
TimeStep = 2, Start time = 0 + 1 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 1.5 (time), but prev timestep's dt
was 1 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.1 (time).
TimeStep = 3, Start time = 1 + 1.1 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 1.65 (time), but prev timestep's dt
was 1.1 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.21 (time).
TimeStep = 4, Start time = 2.1 + 1.21 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 1.815 (time), but prev timestep's dt
was 1.21 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.331 (time).
TimeStep = 5, Start time = 3.31 + 1.331 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 1.9965 (time), but prev timestep's dt
was 1.331 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.4641 (time).
TimeStep = 6, Start time = 4.641 + 1.4641 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 2.19615 (time), but prev timestep's dt
was 1.4641 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.61051 (time).
TimeStep = 7, Start time = 6.1051 + 1.61051 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 2.41577 (time), but prev timestep's dt
was 1.61051 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.77156 (time).
TimeStep = 8, Start time = 7.71561 + 1.77156 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 2.65734 (time), but prev timestep's dt
was 1.77156 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 1.94872 (time).
TimeStep = 9, Start time = 9.48717 + 1.94872 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 2.92308 (time), but prev timestep's dt
was 1.94872 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 2.14359 (time).
TimeStep = 10, Start time = 11.4359 + 2.14359 prev timeStep dt
In FiniteElementContext_CalcNewDt(): dt calculated was 3.21538 (time), but prev timestep's dt
was 2.14359 (time) and max allowed increase percentage is 10.00
, thus limiting current
dt to 2.35795 (time).

Turning on timestep braking, at 80% level, running again - expect same as original:
Run until 10 timeSteps have been run
TimeStep = 1, Start time = 0 + 0 prev timeStep dt
TimeStep = 2, Start time = 0 + 1 prev timeStep dt
TimeStep = 3, Start time = 1 + 1.5 prev timeStep dt
TimeStep = 4, Start time = 2.5 + 2.25 prev timeStep dt
TimeStep = 5, Start time = 4.75 + 3.375 prev timeStep dt
TimeStep = 6, Start time = 8.125 + 5.0625 prev timeStep dt
TimeStep = 7, Start time = 13.1875 + 7.59375 prev timeStep dt
TimeStep = 8, Start time = 20.7812 + 11.3906 prev timeStep dt
TimeStep = 9, Start time = 32.1719 + 17.0859 prev timeStep dt
TimeStep = 10, Start time = 49.2578 + 25.6289 prev timeStep dt
